File: src/com/mistraltech/smogen/codegenerator/GeneratorProperties.java
package com.mistraltech.smogen.codegenerator;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiDirectory;

public abstract class GeneratorProperties<T extends GeneratorProperties> {
    private Project project;
    private String packageName;
    private VirtualFile sourceRoot;
    private String className;
    private PsiDirectory parentDirectory;

    @SuppressWarnings("unchecked")
    protected T self() {
        return (T) this;
    }

    public String getClassName() {
        return className;
    }

    public T setClassName(String className) {
        this.className = className;
        return self();
    }

    public String getPackageName() {
        return packageName;
    }

    public T setPackageName(String packageName) {
        this.packageName = packageName;
        return self();
    }

    public VirtualFile getSourceRoot() {
        return sourceRoot;
    }

    public T setSourceRoot(VirtualFile sourceRoot) {
        this.sourceRoot = sourceRoot;
        return self();
    }

    public PsiDirectory getParentDirectory() {
        return parentDirectory;
    }

    @SuppressWarnings("WeakerAccess")
    public T setParentDirectory(PsiDirectory parentDirectory) {
        this.parentDirectory = parentDirectory;
        return self();
    }

    public Project getProject() {
        return project;
    }

    public T setProject(Project project) {
        this.project = project;
        return self();
    }

    public abstract String getFileName();

    public abstract CodeWriter getCodeWriter();
}

File: src/main/kotlin/com/mistraltech/smogen/codegenerator/GeneratorProperties.kt
package com.mistraltech.smogen.codegenerator

import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.PsiDirectory

abstract class GeneratorProperties<T : GeneratorProperties<T>> {
    var project: Project? = null
        private set
    var packageName: String? = null
        private set
    var sourceRoot: VirtualFile? = null
        private set
    var className: String? = null
        private set
    var parentDirectory: PsiDirectory? = null
        private set

    @Suppress("UNCHECKED_CAST")
    protected fun self(): T {
        return this as T
    }

    fun setClassName(className: String?): T {
        this.className = className
        return self()
    }

    fun setPackageName(packageName: String?): T {
        this.packageName = packageName
        return self()
    }

    fun setSourceRoot(sourceRoot: VirtualFile?): T {
        this.sourceRoot = sourceRoot
        return self()
    }

    fun setParentDirectory(parentDirectory: PsiDirectory?): T {
        this.parentDirectory = parentDirectory
        return self()
    }

    fun setProject(project: Project?): T {
        this.project = project
        return self()
    }

    abstract val fileName: String
    abstract val codeWriter: CodeWriter
}
